### PR TITLE
Branch for working on rust support #12

### PR DESCRIPTION
**Pull request by WaDelma**
Addresses issues #21 and #20.
Implements exercise scanning and points attribute for rust tests via compiler plug-in:
https://github.com/WaDelma/points

### CODE REVIEW

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/RustPointsParser.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Scanner;
+
+public class RustPointsParser {
+
+    /**
+     * Parses a file containing information of points of exercise.
+     */
+    public Optional<ExerciseDesc> parse(Scanner scanner, String exerciseName) {
+        Map<String, List<String>> map = new HashMap<>();
+        while (scanner.hasNextLine()) {
```
Comment (ljleppan) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/RustPointsParser.java]: Consider replacing with [`java.nio.file.Files#readAllLines(Path path, Charset cs)`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#readAllLines(java.nio.file.Path,%20java.nio.charset.Charset)), unless we except the file to be large enough to cause memory issues.

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/RustPointsParser.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            String[] keyValue = scanner.nextLine().split(" = ");
+            if (keyValue.length == 1) {
```
Comment (ljleppan) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/RustPointsParser.java]: Perhaps silly, but is it possible for the input to be malformed in a way that we'd get a 3-way (or larger) split, and would that cause issues further down the road? Essentially, is there any reason to not just have `if (keyValue.length != 2)` here?

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java [hunk 3]
```diff
@@ -61,7 +63,7 @@ protected StudentFilePolicy getStudentFilePolicy(Path projectPath) {
     public ValidationResult checkCodeStyle(Path path) {
         if (run(new String[]{"cargo", "clean"}, path).isPresent()) {
             String[] command = {"cargo", "rustc", "--", "--forbid", "warnings"};
-            log.info("Building for lints with command {0}", Arrays.deepToString(command));
+            log.info("Building for lints with command {}", Arrays.deepToString(command));
```
Comment (jamox) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: Nit: I'm wondering if there is need for the `deepToString` though I don't know whether it has any efficiency implications... 
I'm not saying we need to change it, I just keep on wondering about this.

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java [hunk 4]
```diff
@@ -81,7 +83,25 @@ public String getPluginName() {
...
     public Optional<ExerciseDesc> scanExercise(Path path, String exerciseName) {
-        throw new UnsupportedOperationException("Not supported yet.");
+        if (!isExerciseTypeCorrect(path)) {
+            log.error("Failed to scan exercise due to missing Cargo.toml.");
+            return Optional.absent();
+        }
+
+        try {
+            runTests(path);
+        } catch (Exception e) {
```
Comment (jamox) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: Not really sure whether to catch Exception...

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java [hunk 5]
```diff
@@ -96,7 +116,7 @@ public RunResult runTests(Path path) {
 
     private Optional<RunResult> build(Path path) {
         String[] command = {"cargo", "test", "--no-run"};
-        log.info("Building project with command {0}", Arrays.deepToString(command));
+        log.info("Building project with command {}", Arrays.deepToString(command));
```
Comment (jamox) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: I'd say this won't log the command, just `{}`

Comment (jamox) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: Or maybe I'm just mistaken?

Comment (WaDelma) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: Well if I had {0} it shows as '{0}' in logs and with {} it worked.

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java (hunk 4 above)
Comment (WaDelma) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: This is essentially same as in C make plugin.

File: tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java (hunk 3 above)
Comment (WaDelma) [inline tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java]: This is essentially same as in C make plugin.

**Pull request merged**